File: src/AnalyticsService/Service/Queries/Handlers/GetGenericStatsQueryHandler.cs
using System.Data;
using AnalyticsService.Database.Api;
using AnalyticsService.Service.Model;
using MediatR;

namespace AnalyticsService.Service.Queries.Handlers;

/// <summary>
/// A handler for retrieving generic stats about all document batches in a specified workflow.
/// </summary>
internal sealed class GetGenericStatsQueryHandler(
    IDbConnection conn,
    IBatchStatRepository batchStatRepository)
    : IRequestHandler<GetGenericStatsQuery, GeneralBatchStat>
{
    /// <inheritdoc/>
    public async Task<GeneralBatchStat> Handle(GetGenericStatsQuery request, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
            return GeneralBatchStat.Default;
        var stats = (await batchStatRepository
            .GetBatchStats(conn, request.WorkflowId))
            .ToList();
        if (stats.Count == 0)
            return GeneralBatchStat.Default;

        var successRate = CalculateSuccessRate(
            stats.Count,
            stats.Count(i => i.IsSuccess())
        );
        return new GeneralBatchStat(
            (int)stats.Average(s => s.NumberOfDocuments),
            successRate,
            TimeSpan.FromTicks((long)stats.Average(d => d.RunTime.Ticks))
        );
    }

    /// <summary>
    /// Method for calculating a success rate of document batches.
    /// </summary>
    private static float CalculateSuccessRate(int all, int successful)
        => all / (float)successful;
}

File: src/AnalyticsService/Service/Queries/Handlers/GetGenericStatsQueryHandler.fs
namespace AnalyticsService.Service.Queries.Handlers

open System
open System.Data
open AnalyticsService.Service.Commands
open AnalyticsService.Service.Model
open AnalyticsService.Database.Repositories
open MediatR

/// A handler for retrieving generic stats about all document batches in a specified workflow.
[<Sealed>]
type internal GetGenericStatsQueryHandler(
    conn: IDbConnection,
    batchRepo: IBatchStatRepository
) =
    let calculateSuccessRate all successful = all / float(successful)

    interface IRequestHandler<GetGenericStatsQuery, GeneralBatchStats> with
        member this.Handle(request, cancellationToken) =
            task {
                if not(cancellationToken.IsCancellationRequested) then
                    let stats =
                        batchRepo.GetBatchStats request.WorkflowId conn
                        |> Async.AwaitTask
                        |> Async.RunSynchronously
                        |> Seq.toList

                    if stats.Length = 0 then
                        return GeneralBatchStats.Default
                    else
                        return {
                            AvgNumberOfDocs =
                                stats
                                |> List.averageBy(fun i -> float i.NumberOfDocuments)
                                |> Math.Round
                                |> int
                            SuccessRate =
                                stats
                                |> List.filter(fun i -> i.IsSuccess)
                                |> List.length
                                |> calculateSuccessRate stats.Length
                            AvgDuration =
                                stats
                                |> List.averageBy(fun i -> float i.RunTime.Ticks)
                                |> int64
                                |> TimeSpan.FromTicks
                        }
                else
                    return GeneralBatchStats.Default
            }